### DISCUSSION TITLE
Reduce complexity of the Docker Image

### DISCUSSION THREAD

**Discussion by timlevett** [grafana/grafana#54165] [Platform]
Right now Grafana has three `Dockerfile`s, one in `/` and two others in `/packaging/docker`. The one in the root directory seems somewhat unmaintained based on recent breakages and is also not part of the CI pipeline. It also builds the binary and then generates the running image which is not typical for Docker images. The ones in packaging a required bundle of a Grafana build that is not straightforward to how to generate those bundles. 

With that in mind, I'd like to start a discussion to come up with a way of easily generating a docker image of Grafana from source.

Some ideas (more are welcome!):
* Keep the /packaging/docker but provide a make command that generates the bundle it requires. 
* Reduce the need for /packaging/docker from needing a bundle to just a binary. 
* Start maintaining the `Dockerfile` in `/` by adding it to the CI pipeline and also start using it as the release artifact generator.

**Comment by guicaulada** [comment]
I think keeping the one in packaging and deprecating the root Dockerfile would be the best approach, however right now to build a container from source there's a script that is not part of the repository. I think at first, if possible, we should make this script available and iterate over it producing a more robust version of the script that can support all systems and use cases, eventually adding it to the Makefile as an option to make it as easy as possible to generate a container from source.

**Reply by kminehart** [reply]
I think I disagree with this: "I think keeping the one in packaging and deprecating the root Dockerfile would be the best approach"

That Dockerfile is particularly tricky and really not conventionally Docker-esque. It takes an already packaged version of Grafana and creates a docker image from it, and sort of subverts expectations that everyone has about how Dockerfiles work and where they should live in a repository.

99% of the time, if a project has a `Dockerfile`, it can be built with `docker build`.

I do agree that we only need one Dockerfile. I think the existing one in the root of the project, however, would be the right one to start with. A lot of people expect to be able to use Docker to develop Grafana, and keeping it at the root of the project would encourage that.

**Reply by kminehart** [reply]
There's also this: https://github.com/grafana/grafana/issues/24134 :see_no_evil:

**Reply by guicaulada** [reply]
I understand. I agree having one in the root of the project and being able to build with `docker build` feels more common and Docker-esque, but as a quicker solution so our partners can build their own image I feel like cleaning and making the script that's being used internally available in the packaging folder might be the fastest route.

But I agree, I would really appreciate being able to run `docker build` and build from source with no problems.

**Reply by kminehart** [reply]
> as a quicker solution so our partners can build their own image I feel like cleaning and making the script that's being used internally available in the packaging folder might be the fastest route

Oh! absolutely. I provided them a script that they should be able to use to get that docker image built. Whether or not I can commit to the repository is probably a difficult question to answer, it seems like in the past we've purposefully tried to omit stuff like that.

After starting some development on this, I see why we have the Dockerfile that builds Grafana based on an image. Compiling the javascript and Go in the `docker build` process is abysmally slow.

**Reply by andresmgot** [reply]
I don't have much context about the building environment but my +1 to have a standalone Dockerfile in the root of the repo (or somewhere else) which builds everything and generate the final image (and we use it on our release process too).

> Compiling the javascript and Go in the `docker build` process is abysmally slow.

Using docker should be the same than in a fresh linux environment. Being slow should mean that it's not using any cached stuff, which helps with the reproducibility of the builds. For a dev environment there are always solutions to mount Go or yarn cache to speed up builds.

**Reply by kminehart** [reply]
I 100% agree with you.

I should have clarified one thing: since this docker image is used in CI, allowing it to use an existing built frontend / backend can save us a ton of time in the CI process.

When using it locally it totally should be from scratch without any existing assets.

**Reply by xlson** [reply]
The Dockerfile in root and the ones in packaging have very different goals and to me that makes sense. But obviously it creates a maintenance burden for us, especially since the one in root is not a part of the CI builds, meaning that it can break at any time.

- the ones in packaging is there to create docker images for specific releases of Grafana, as such I would argue that it makes sense that it uses a build of Grafana rather than building Grafana from scratch. We could possibly make it so that it downloads a package if no package is specified so that it always does something?
- the one in root is meant to support developers that want a docker image running their current grafana workspace contents. It duplicates our whole build process without using our tooling which obviously is no small feat and also quite cumbersome and it breaks easily. I think it makes sense to have this image since it makes docker the only prerequisite to be able to build and run grafana. But it should use our build tools and build images to simplify the process (I'm taking about the process of maintaining the Dockerfile) and make it more similar to what we do in ci.

> Using docker should be the same than in a fresh linux environment. Being slow should mean that it's not using any cached stuff, which helps with the reproducibility of the builds. For a dev environment there are always solutions to mount Go or yarn cache to speed up builds.

Unfortunately I don't think mounting the build caches work a lot of the time. If you run amd64/linux it should work, but a lot of devs are on darwin, darwin m1 or even windows.


So where am I going with these comments. My point is that it makes sense to have both kind of images but that if we want to keep all of them around we need to start maintaining the one in root, add it to ci builds and improve it. I'm also very skeptical of building Grafana from scratch for the packaging docker image, it seems really strange to me go through the whole build process again. I question the statement that it's best practice to build everything from scratch in the docker images. Perhaps my docker knowledge is out-of-date but I've seen plenty of official images that use pre-puilt debian packages or tar-balls to generate their docker images.

**Reply by kminehart** [reply]
@xlson well said, I agree. I wonder if there's a way to do what you're suggesting and combine the two. That would at least reduce the amount of Dockerfiles that have to be maintained without making the CI process take extra time to verify a non-pertinent Dockerfile.

The most important thing for me is that the Dockerfile that is used for development produces an image that is as similar as possible to what is available in Docker Hub.

**Reply by xlson** [reply]
One possible way to simplify maintenance could be to use a script to generate the Dockerfile in root and (possibly) the one in packaging as well.

We need two base Dockerfiles that:
- builds Grafana from scratch and produces a tar-ball (or some other artifact that we can use in the packaging image)
- the packaging image that takes a tar-ball and produces a release image

Then we would need a script that generates:
- the packaging dockerfile (this step might not be necessary if we can just use this dockerfile as it is)
- the root dockerfile which would be a combination of the one that builds grafana from scratch and the packaging dockerfile

If we had something like that we'd just re-generate the generated ones whenever we make changes to the templated ones and we wouldn't need to maintain two different version of the packaging ones. We should also be able to setup a nightly build for the one that builds grafana from scratch, we could also run that build step anytime it is modified.  

Other than that I would have that we could simplify work on the image that builds grafana from scratch by using grabpl or anything that replaces it and base those steps on our official build images.

**Comment by GuYounes** [comment]
Any news about this ? 
I don't have a lot of knowledge about grafana workflow etc, but I can give a user point of view.

It would be nice to have a documentation with the steps to make a "production build" that would be a docker image or a windows installer etc. We can see a lot of scripts in the project but it seems like none of them would do that. Same for the root Dockerfile which seems to be out of date and not even running without issues. I understand better with this discussion why.

`I think at first, if possible, we should make this script available and iterate over it producing a more robust version of the script that can support all systems and use cases, eventually adding it to the Makefile as an option to make it as easy as possible to generate a container from source.`

Is there any constraint to put this script available in the repository ? I think it would be really helpfull.
